include/sleip/dynamic_array.hpp

```diff
@@ -0,0 +1,320 @@
+#ifndef SLEIP_DYNAMIC_ARRAY_HPP_
+#define SLEIP_DYNAMIC_ARRAY_HPP_
+
+#include <boost/core/alloc_construct.hpp>
+#include <boost/throw_exception.hpp>
+#include <boost/config.hpp>
+
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <utility>
+#include <type_traits>
+
+namespace sleip
+{
+template <class T, class Allocator>
+struct dynamic_array;
+
+template <class T, class Allocator = std::allocator<T>>
+struct dynamic_array
+{
+public:
+  using value_type             = T;
+  using allocator_type         = Allocator;
+  using size_type              = std::size_t;
+  using difference_type        = std::ptrdiff_t;
+  using reference              = value_type&;
+  using const_reference        = value_type const&;
+  using pointer                = value_type*;
+  using const_pointer          = value_type const*;
+  using iterator               = pointer;
+  using const_iterator         = const_pointer;
+  using reverse_iterator       = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+private:
+  T*          data_ = nullptr;
+  std::size_t size_ = 0;
+  Allocator   alloc_;
+
+public:
+  dynamic_array() noexcept(noexcept(Allocator())) = default;
+
+  explicit dynamic_array(const Allocator& alloc) noexcept
+    : alloc_(alloc)
+  {
+  }
+
+  dynamic_array(size_type count, T const& value, Allocator const& alloc = Allocator())
+    : alloc_(alloc)
+  {
+    data_ = std::allocator_traits<Allocator>::allocate(alloc_, count);
+
+#ifdef BOOST_NO_EXCEPTIONS
```

Comment:
I suggest guarding the `try` and `catch` parts, instead of duplicating the implementation for the happy path.